### PR TITLE
fix: Lock functions from building with same codeuri at the same time

### PR DESCRIPTION
**Pull request by mildaniel**
#### Which issue(s) does this change fix?

#3587

#### Why is this change necessary?
Resolves a race condition where functions with the same codeuri attempt to modify shared resources simultaneously during a build.

#### How does it address the issue?
Adds another lock to a function whenever functions share the same codeuri.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by mndeveci** [comment]
I think we don't need following folders in the repository since they will be created once the function is built;
- `tests/integration/testdata/sync/code/after/dotnet_function/src/HelloWorld/bin/`
- `tests/integration/testdata/sync/code/before/dotnet_function/src/HelloWorld/bin/`

Rest looks good to me 👍🏼

File: tests/integration/testdata/sync/code/before/template-python.yaml [hunk 1]
```diff
@@ -16,6 +16,24 @@ Resources:
         - Ref: HelloWorldLayer
```
Comment (mndeveci) [inline tests/integration/testdata/sync/code/before/template-python.yaml:16]: This file name is no longer valid anymore (`template-python.yaml`). Let's move this function into its own template.

File: tests/integration/testdata/sync/code/before/dotnet_function/src/HelloWorld/obj/HelloWorld.csproj.nuget.dgspec.json [hunk 2]
```diff
@@ -0,0 +1,81 @@
+{
```
Comment (mndeveci) [inline tests/integration/testdata/sync/code/before/dotnet_function/src/HelloWorld/obj/HelloWorld.csproj.nuget.dgspec.json:1]: Sorry if I missed this my earlier pass, but it looks like we also don't need this `obj` folder under function source. I believe this is generated during build so we don't need to push this in git.

File: tests/integration/testdata/sync/code/after/dotnet_function/src/HelloWorld/obj/HelloWorld.csproj.nuget.dgspec.json [hunk 3]
```diff
@@ -0,0 +1,81 @@
+{
```
Comment (mndeveci) [inline tests/integration/testdata/sync/code/after/dotnet_function/src/HelloWorld/obj/HelloWorld.csproj.nuget.dgspec.json:1]: Same for this `obj` folder as well, we probably don't need this.

**Review by mndeveci — CHANGES_REQUESTED** [review]

**Review by mndeveci — APPROVED** [review]
LGTM!

**Review by qingchm — APPROVED** [review]

**Pull request merged**